Allow defaultTableOptions to be overridden from configuration

ruifung opened the pull request:
In my installation, MariaDB is configured to enable page compression by default while supporting utf8mb4.
The default table creation options in that case causes issue as the COMPRESSED row format is not compatible with MariaDB page compression.

Therefore, this allows the default table creation options to be overridden from the configuration if required and will take precedence over the defaults where mysql.utf8m4 is true. If not specified, it should behave exactly as it was.

That being said, I think that it shouldn't be defaulting to compressed especially since according to (https://mariadb.com/kb/en/innodb-compressed-row-format/) 

> From MariaDB 10.6.0, tables that are of the COMPRESSED row format are read-only by default. This is the first step towards removing write support and deprecating the feature.


Closes #25436

ruifung committed:
Allow defaultTableOptions to be overridden from configuration.

Signed-off-by: Yip Rui Fung <rf@yrf.me>

kesselb commented:
```
if ($this->config->getValue('mysql.utf8mb4', false)) {
	$connectionParams['defaultTableOptions'] = [
		'collate' => 'utf8mb4_bin',
		'charset' => 'utf8mb4',
		'row_format' => $this->config->getValue('mysql.utf8mb4_row_format', 'compressed'),
		'tablePrefix' => $connectionParams['tablePrefix']
	];
}
```

I had a simpler version in mind ;) But this might work as well.

ruifung commented:
> 
> 
> ```
> if ($this->config->getValue('mysql.utf8mb4', false)) {
> 	$connectionParams['defaultTableOptions'] = [
> 		'collate' => 'utf8mb4_bin',
> 		'charset' => 'utf8mb4',
> 		'row_format' => $this->config->getValue('mysql.utf8mb4_row_format', 'compressed'),
> 		'tablePrefix' => $connectionParams['tablePrefix']
> 	];
> }
> ```
> 
> I had a simpler version in mind ;) But this might work as well.

I figured that rather then just overriding one parameter, it might be more useful to be able to override the entire set of table creation options (like say, PAGE_COMPRESSED=1, for example.)

kesselb commented:
Useful != Good. If it's possible to change collate or charset then people may change it. For example set it to latin2 for whatever reason.

ruifung commented:
> 
> 
> Useful != Good. If it's possible to change collate or charset then people may change it. For example set it to latin2 for whatever reason.

That... could pose some problems... hmm... not quite sure how this would work to limit uh... user error while allowing for flexibility with... nonstandard DB configurations?

Unless maybe the collate / charset is applied later to override the values from config? I do agree it doesn't make much sense to be change out the collation / charset beyond what is already provided though.

ruifung committed:
Prevent user from being able to override the collate / charset table creation options.

Bad things would probably happen if they changed it to something that isn't some kind of unicode encoding.

Signed-off-by: Yip Rui Fung <rf@yrf.me>

nickvergessen commented:
Maybe we can list some allowed parameters instead and then filter others out?
Also we are trying to leak less doctrine on our database API levels. adding direct array with an unfiltered list of doctrine options is somewhat the opposite.

ruifung commented:
@nickvergessen If I understand correctly, you are suggesting to add a whitelist of allowed options for this?

Something like $allowed_options = ['row_format', 'page_compressed'] and so on?

In that case I think I would need a little assistance figuring out what options to include there... given that I have little experience with the other databases this could potentially be deployed with.

nickvergessen commented:
The only once I can see being used are:
![Bildschirmfoto von 2021-02-10 11-29-35](https://user-images.githubusercontent.com/213943/107497876-5f68cb80-6b93-11eb-9a04-e52f597f07ed.png)

GieltjE commented:
This fix should get some priority, as MariaDB 10.6.3 has been declared stable and marks tables with COMPRESSION as read only: https://mariadb.com/docs/reference/mdb/system-variables/innodb_read_only_compressed/

juliushaertl commented:
Considering the quote from the mariadb docs I'm wondering if we actually should drop the compressed and instead recommend to use page compression.

https://mariadb.com/kb/en/innodb-compressed-row-format/

> In MariaDB 10.1 and later, an alternative (and usually superior) way to compress InnoDB tables is by using InnoDB Page Compression. See Comparison with the COMPRESSED Row Format.

> This is the first step towards removing write support and deprecating the feature.

GieltjE commented:
@juliushaertl Agree, also given that MySQL 5.6 (2013) supports DYNAMIC row format and 5.7 (2015)  supports page compression the situation doesn't seem to differ between MariaDB and MySQL.
We might want to consider making DYNAMIC row format the default.
